# **PROJECT 1: Customer Churn Analysis - Telecom Industry** #

## **Columns** ##

Customer churn is a major issue in telecommunications, where retaining customers is significantly more cost-effective than acquiring new ones. \
This dataset contains customer-level details including demographics, services subscribed, billing information, and whether the customer churned.

**CustomerID**: Unique identifier

**Gender**: Male/Female

**SeniorCitizen**: 0 = No, 1 = Yes

**Partner**: Yes/No

**Dependents**: Yes/No

**Tenure**: Number of months with the company

**PhoneService**: Yes/No

**MultipleLines**: Yes/No/No phone service

**InternetService**: DSL/Fiber optic/No

**OnlineSecurity**, **OnlineBackup**, **DeviceProtection**, **TechSupport**, **StreamingTV**, **StreamingMovies**: Yes/No/No internet service

**Contract**: Month-to-month/One year/Two year

**PaperlessBilling**: Yes/No

**PaymentMethod**: Electronic check/Mailed check/Bank transfer/Credit card

**MonthlyCharges**: Amount billed monthly

**TotalCharges**: Lifetime total charges

**Churn**: Yes/No


In [146]:
!python.exe -m pip install --upgrade pip
!pip install scikit-learn

Defaulting to user installation because normal site-packages is not writeable
  Using cached pip-25.3-py3-none-any.whl.metadata (4.7 kB)
Using cached pip-25.3-py3-none-any.whl (1.8 MB)


  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.


Defaulting to user installation because normal site-packages is not writeable
  Using cached threadpoolctl-3.6.0-py3-none-any.whl.metadata (13 kB)
   ---------------------------------------- 0.0/8.0 MB ? eta -:--:--
   ---------- ----------------------------- 2.1/8.0 MB 15.6 MB/s eta 0:00:01
   ---------------------------------------- 8.0/8.0 MB 25.3 MB/s  0:00:00
   ---------------------------------------- 0.0/36.3 MB ? eta -:--:--
   -------- ------------------------------- 8.1/36.3 MB 41.4 MB/s eta 0:00:01
   ------------------ --------------------- 16.8/36.3 MB 40.8 MB/s eta 0:00:01
   --------------------------- ------------ 25.2/36.3 MB 40.7 MB/s eta 0:00:01
   ------------------------------------ --- 33.6/36.3 MB 40.1 MB/s eta 0:00:01
   ---------------------------------------- 36.3/36.3 MB 35.9 MB/s  0:00:01
Using cached threadpoolctl-3.6.0-py3-none-any.whl (18 kB)

   ---------- ----------------------------- 1/4 [scipy]
   ---------- ----------------------------- 1/4 [scipy]
 

In [25]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt 
import seaborn as sn 
import warnings
warnings.filterwarnings('ignore')
%matplotlib inline

In [26]:
# Load CSV file
telecom = pd.read_csv('../data/raw/telecom_customer_churn.csv')
telecom.head(10)

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes
5,9305-CDSKC,Female,0,No,No,8,Yes,Yes,Fiber optic,No,...,Yes,No,Yes,Yes,Month-to-month,Yes,Electronic check,99.65,820.5,Yes
6,1452-KIOVK,Male,0,No,Yes,22,Yes,Yes,Fiber optic,No,...,No,No,Yes,No,Month-to-month,Yes,Credit card (automatic),89.10,1949.4,No
7,6713-OKOMC,Female,0,No,No,10,No,No phone service,DSL,Yes,...,No,No,No,No,Month-to-month,No,Mailed check,29.75,301.9,No
8,7892-POOKP,Female,0,Yes,No,28,Yes,Yes,Fiber optic,No,...,Yes,Yes,Yes,Yes,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes
9,6388-TABGU,Male,0,No,Yes,62,Yes,No,DSL,Yes,...,No,No,No,No,One year,No,Bank transfer (automatic),56.15,3487.95,No


In [27]:
# Find out how many columns and rows of data is in the file 
telecom.shape

(7043, 21)

In [28]:
# Get statistical information
telecom.describe()

,SeniorCitizen,tenure,MonthlyCharges
count,7043.000000,7043.000000,7043.000000
mean,0.162147,32.371149,64.761692
std,0.368612,24.559481,30.090047
min,0.000000,0.000000,18.250000
25%,0.000000,9.000000,35.500000
50%,0.000000,29.000000,70.350000
75%,0.000000,55.000000,89.850000
max,1.000000,72.000000,118.750000


***So far, we know there are 21 columns and 7043 rows in the dataframe; and only SeniorCitizen, tenure, and MonthlyCharges are numbers, the rest is catagorical data.*** \
***Now, we will look at the data type of each column and whether there are null values***

In [29]:
telecom.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [30]:
telecom.isna().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [31]:
telecom['TotalCharges'][telecom['TotalCharges'].isnull()]

Series([], Name: TotalCharges, dtype: object)

***There are 11 null values in the dataset and they are all related to TotalCharges.*** \
***Those null TotalCharges values probably pointing to records for new customers.*** \
***We can either delete them or fill them with 0.*** \
***I chose to fill them with 0 because we don't want to neglect any data for the analysis especially those that allow us to tell a well-rounded story.***\
***Before doing that, we have to change the data type of TotalCharges from object to float64***





In [32]:
telecom['TotalCharges'] = pd.to_numeric(telecom['TotalCharges'], errors='coerce')

In [33]:
telecom['TotalCharges'].dtype

dtype('float64')

In [34]:
telecom['TotalCharges'].fillna(0, inplace=True)

In [35]:
telecom.isna().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [36]:
telecom.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


***Looks like the dataset doesn't have any null values in there anymore.*** \
***And it is confirmed that the dataset only have 4 numeric columns, and the rest are categorical data.*** \
***Then, I would like to make sure that the categorical data does not contain inconsistency and error.***

***We can start with checking the unique values in each categorical column.***

In [37]:
categorical_cols = telecom.select_dtypes(include=['object']).columns
print(f"\nUnique values in categorical columns:\n")
for col in categorical_cols:
    if col != 'customerID':
        print(f"{col}: {telecom[col].unique()[:5]}... ({len(telecom[col].unique())} unique values)")


Unique values in categorical columns:

gender: ['Female' 'Male']... (2 unique values)
Partner: ['Yes' 'No']... (2 unique values)
Dependents: ['No' 'Yes']... (2 unique values)
PhoneService: ['No' 'Yes']... (2 unique values)
MultipleLines: ['No phone service' 'No' 'Yes']... (3 unique values)
InternetService: ['DSL' 'Fiber optic' 'No']... (3 unique values)
OnlineSecurity: ['No' 'Yes' 'No internet service']... (3 unique values)
OnlineBackup: ['Yes' 'No' 'No internet service']... (3 unique values)
DeviceProtection: ['No' 'Yes' 'No internet service']... (3 unique values)
TechSupport: ['No' 'Yes' 'No internet service']... (3 unique values)
StreamingTV: ['No' 'Yes' 'No internet service']... (3 unique values)
StreamingMovies: ['No' 'Yes' 'No internet service']... (3 unique values)
Contract: ['Month-to-month' 'One year' 'Two year']... (3 unique values)
PaperlessBilling: ['Yes' 'No']... (2 unique values)
PaymentMethod: ['Electronic check' 'Mailed check' 'Bank transfer (automatic)'
 'Credit card 

***The values seem to look okay. Some of the fields, e.g., MultipleLines, includes a value 'No' and a value 'No phone service'.*** \
***I'm debating whether I should combine those two values into one , 'No', only.*** \
***I think I'm going to proceed without merging the data for now for the sake of flexibility.***

***Then I would like to create age groups for the tenure data.***

In [38]:
telecom['TenureGroup'] = pd.cut(telecom['tenure'], 
                           bins=[0, 12, 24, 36, 48, 60, 72, 84, 96, 108, 120],
                           labels=['0-1yr', '1-2yr', '2-3yr', '3-4yr', '4-5yr', 
                                  '5-6yr', '6-7yr', '7-8yr', '8-9yr', '9-10yr'])

In [39]:
telecom.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn,TenureGroup
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No,0-1yr
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,No,No,No,One year,No,Mailed check,56.95,1889.50,No,2-3yr
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,0-1yr
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No,3-4yr
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,0-1yr


***The data seems to look okay.***

***At this point, it might be helpful if we can figure out the average total charge of services per month.***

In [138]:
def filter_out_no_service(service):
    # Convert dataframe object into series object
    service_series = service.squeeze()
    
    # Convert to string dtype
    service_str = service_series.astype('string')
    
    # Create boolean mask: True if NOT starting with 'No'
    a = ~ service_str.str.startswith('No')
    # print(a)
    
    return a

In [139]:
service_columns = ['PhoneService', 'InternetService', 'StreamingTV', 'StreamingMovies']
telecom['ServicesCount'] = telecom[service_columns].apply(lambda x: filter_out_no_service(x).sum(), axis=1)
telecom['AvgChargePerService'] = telecom['MonthlyCharges'] / telecom['ServicesCount'].replace(0, 1)

In [140]:
telecom.head(6)

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn,TenureGroup,ServicesCount,AvgChargePerService
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,Month-to-month,Yes,Electronic check,29.85,29.85,No,0-1yr,1,29.8500
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,No,One year,No,Mailed check,56.95,1889.50,No,2-3yr,2,28.4750
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,0-1yr,2,26.9250
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,No,One year,No,Bank transfer (automatic),42.30,1840.75,No,3-4yr,1,42.3000
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,0-1yr,2,35.3500
5,9305-CDSKC,Female,0,No,No,8,Yes,Yes,Fiber optic,No,...,Yes,Month-to-month,Yes,Electronic check,99.65,820.50,Yes,0-1yr,4,24.9125


In [153]:
telecom.dtypes

customerID               object
gender                   object
SeniorCitizen             int64
Partner                  object
Dependents               object
tenure                    int64
PhoneService             object
MultipleLines            object
InternetService          object
OnlineSecurity           object
OnlineBackup             object
DeviceProtection         object
TechSupport              object
StreamingTV              object
StreamingMovies          object
Contract                 object
PaperlessBilling         object
PaymentMethod            object
MonthlyCharges          float64
TotalCharges            float64
Churn                    object
TenureGroup            category
ServicesCount             int64
AvgChargePerService     float64
dtype: object

***In order to find out the correlation, we need to convert the categorical columns into nunerical columns***

In [148]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, StandardScaler

telecom_encoded = telecom.copy()

le = LabelEncoder()

for col in telecom_encoded.select_dtypes(include=['object']).columns:
    # I don't think we need to covert customer ID into numerical form,
    # and churn is a value we are looking for and should be treated as its own label; 
    # so let's single both columns out from the Dataframe
    if col not in ['customerID', 'Churn']:
        telecom_encoded[col] = le.fit_transform(telecom_encoded[col].astype('string'))

telecom_encoded['Churn'] = le.fit_transform(telecom_encoded['Churn'])

In [150]:
# Calculate correlations
correlations = telecom_encoded.corr()['Churn'].sort_values(ascending=False)

ValueError: could not convert string to float: '0-1yr'

In [149]:
correlations.head(6)

NameError: name 'correlations' is not defined

***Let's check the dimension of our DataFrame***

In [141]:
telecom.shape

(7043, 24)